### DISCUSSION TITLE
Importing type variants (continued from discussion #3232)

### DISCUSSION THREAD

**Discussion by barcek** [gleam-lang/gleam#3512] [Ideas & suggestions]
Adding a thought on this subject, re something I've noticed early on using Gleam. Importing a larger number of variants can get unwieldy, especially having to update imports if the custom type is extended or reduced at a later point in development.

From @Dan-C-Reed's use in [discussion #3232](https://github.com/gleam-lang/gleam/discussions/3232)):

```gleam
import types.{type DataSource, Url, File}
```

This works, but is just two variants. There could be more:

```gleam
import types.{type DataSource, Url, File, Record, Commit, ...}
```

It would be useful if all variants didn't need to be imported, but could be accessed via the type in place, e.g.:

```gleam
import types.{type DataSource}
...
  "url" -> DataSource.Url
...
```

This would help clarify the relationship between variants locally too.

Using the qualified import syntax, i.e. accessing via the module name, is possible and reasonable if there's only one custom type per module, but with multiple types each with multiple variants also potentially confusing.

**Comment by lpil** [comment]
Types are not namespaces, modules are. Your code would be written like so:

```gleam
import types
...
  "url" -> types.Url
...
```

**Reply by barcek** [reply]
Type and module make sense. Using the module name here is tidy for sure, and tidiest with one variant per public type, less so with more variants and more types. Which variants belong to which type could then be less obvious, all else being equal, viewed from within the importing module.

I should clarify I tend to find enums quite useful, and I do often see decently large numbers of variants, plus I tend to see modules formed more around functionality, not necessarily with one or even a few types. No surprise there I guess - there are plenty of language traditions, and we can end up working on quite different codebases even within a single tradition.

This would encourage me to use one module for each multi-variant public type, which isn't an issue in itself. I am finding imports tend to stack up at the top of a file, but that could be how new I am to the language, and I can improve how I organize my Gleam code into modules more generally. There's plenty of nuance I still have to pick up.

Thanks for the reply. It's because Gleam is so good that so many of us are here.
